### ISSUE TITLE
Aborting continuous listening on _global_changes is leaking resources

### ISSUE THREAD

**Issue by redgeoff** [apache/couchdb#1063]
Get ready, this is a fun one! :D

## Overview
- I suspect that there is a resource leak where CouchDB does not properly release all the associated resources when a continuous feed is aborted. The test scripts below continuously update docs while listening on the _changes feed of the _global_changes database. The _global_changes stream is canceled every second.
- This might seem like a rather harmless issue in an isolated test like this, but in a production environment with a fair bit of load, this resource leak can stop your database from scaling as your nodes can easily reach 100% CPU usage and become unresponsive, regardless of the number of nodes you have.
- In  my case, I was able to use feed=longpoll instead of feed=continuous to prevent this resource leak, but this alternate usage may not be a viable option for all use cases. Moreover, if there is a resource leak it could be causing problems in other areas of CouchDB.

## Expected Behavior
CPU load should be distributed equally across nodes in cluster.

## Current Behavior
A single node in the cluster is reaching 100% CPU usage and load is not being distributed equally across nodes in the cluster.

## Possible Solution
I don't currently have a solution, but if your design permits, you can work around this by using feed=longpoll instead of feed=continuous

## Definition of a “workhorse” node
Let’s define the “workhorse node” as the one node in our cluster of nodes that is using a lot of CPU--nearly 100% CPU. This workhorse node will be locked in as the workhorse node until it is restarted.

## Hardware:
- You should be able to replicate the following tests on any hardware, but the CouchDB nodes must be located on different servers. The nodes can be in the same datacenter/AvailabilityZone, but each node MUST be located on a different server. I believe this is because there needs to be some hop distance between the nodes before the problem becomes visible.
- I developed these steps using two AWS t2.nano instances (1 CPU core and 0.5GB memory) running Ubuntu 16. If your hardware is more beefy it is likely that you may need to wait longer or modify the test scripts to pound the CouchDB cluster harder with more concurrency.

## changes-resource-leak-setup.js
```js
var Slouch = require('couch-slouch');
var slouch = new Slouch('http://admin:secret@example.com:5984');

var NUM_DOCS = 100;

var dbName = 'test_db';

var lastSeq = undefined;

var createOrUpdateDocs = function (i) {
  var doc = {
    _id: '' + i,
    foo: 'bar',
    updated_at: (new Date()).toISOString()
  };

  console.log('creating/updating doc=', doc);

  return slouch.doc.createOrUpdate(dbName, doc).then(function () {
    if (i < NUM_DOCS - 1) {
      return createOrUpdateDocs(i + 1);
    }
  });
};

var createDatabaseIfMissing = function (dbName) {
  return slouch.db.exists(dbName).then(function (exists) {
    if (!exists) {
      console.log('creating database', dbName);
      return slouch.db.create(dbName);
    }
  })
};

var createDocs = function () {
  return createOrUpdateDocs(0);
};

createDatabaseIfMissing('_global_changes').then(function () {
  return createDatabaseIfMissing('test_db');
}).then(function () {
  return createDocs();
});
```

## changes-resource-leak-bad-test.js
```js
var Slouch = require('couch-slouch');
var slouch = new Slouch('http://admin:secret@example.com:5984');

var NUM_DOCS = 100;

var dbName = 'test_db';

var lastSeq = undefined;

var createOrUpdateDocs = function (i) {
  var doc = {
    _id: '' + i,
    foo: 'bar',
    updated_at: (new Date()).toISOString()
  };

  // console.log('creating/updating doc=', doc);

  return slouch.doc.createOrUpdate(dbName, doc).then(function () {
    if (i < NUM_DOCS - 1) {
      return createOrUpdateDocs(i + 1);
    }
  });
};

var updateDocs = function () {
  return createOrUpdateDocs(0).then(function () {
    // Repeat
    return updateDocs();
  });
};

var listenToGlobalChanges = function () {
  var first = false;

  var iterator = slouch.db.changes('_global_changes', {
    include_docs: true,
    feed: 'continuous',
    heartbeat: true,
    since: lastSeq,
    limit: 10
  });

  return iterator.each(function (change) {
    console.log('change=', change);
    lastSeq = change.seq;

    if (!first) {
      setTimeout(function () {
        iterator.abort();
      }, 1000);
      first = true;
    }
  }).then(function () {
    // Repeat
    return listenToGlobalChanges();
  });
};

updateDocs();
listenToGlobalChanges();
```

## changes-resource-leak-good-test.js
```js
var Slouch = require('couch-slouch');
var slouch = new Slouch('http://admin:secret@example.com:5984');

var NUM_DOCS = 100;

var dbName = 'test_db';

var lastSeq = undefined;

var createOrUpdateDocs = function (i) {
  var doc = {
    _id: '' + i,
    foo: 'bar',
    updated_at: (new Date()).toISOString()
  };

  // console.log('creating/updating doc=', doc);

  return slouch.doc.createOrUpdate(dbName, doc).then(function () {
    if (i < NUM_DOCS - 1) {
      return createOrUpdateDocs(i + 1);
    }
  });
};

var updateDocs = function () {
  return createOrUpdateDocs(0).then(function () {
    // Repeat
    return updateDocs();
  });
};

var listenToGlobalChanges = function () {
  var iterator = slouch.db.changes('_global_changes', {
    include_docs: true,
    feed: 'longpoll',
    heartbeat: true,
    since: lastSeq,
    limit: 10,
    timeout: 1000
  });

  return iterator.each(function (change) {
    console.log('change=', change);
    lastSeq = change.seq;
  }).then(function () {
    // Repeat
    return listenToGlobalChanges();
  });
};

updateDocs();
listenToGlobalChanges();
```

## Setup:
- Download the changes-resource-leak-setup.js, changes-resource-leak-bad-test.js and changes-resource-leak-good-test.js scripts above the and modify the CouchDB URLs
- Create a 2-node cluster with the default local.ini config. We create a 2-node cluster as it make it easier to see the problem. You will also need to set up a load balancer in front of the 2 nodes, e.g. haproxy, nginx, etc… You do not need to create any system databases as the test script will create _global_changes if the _global_changes database doesn’t exist
- Install node (https://nodejs.org/en/download/package-manager/)
- $ npm install couch-slouch
- $ node ./changes-resource-leak-setup.js
- Wait about 100 secs for 100 docs to be created

## Test 1 - Create a workhorse node:
- $ node ./changes-resource-leak-bad-test.js # This is will run indefinitely
- While changes-resource-leak-bad-test.js runs use top on **both** CouchDB nodes to monitor CPU usage. After about 10 minutes, one of the nodes will reach almost 100% CPU and stay around 100% CPU usage. We’ll refer to this node as the workhorse node. The other node will have some CPU usage, but not nearly as much as the workhorse node.
- Now stop changes-resource-leak-bad-test.js (with Ctrl+C)
- Wait for the CPU usage on both nodes to go down to about 0%
- Now start up ./changes-resource-leak-bad-test.js again
- Within about 20 seconds the workhorse node should again return to almost 100% CPU usage, while the other node will have significantly less usage. This workhorse node will continue to be the workhorse node until you restart your CouchDB nodes.

## Test 2 - Use feed=longpoll instead of feed=continuous
- To illustrate that the problem is most likely due to a resource leak with aborting continuous listening, we’ll run changes-resource-leak-good-test.js, which has roughly the same business logic as changes-resource-leak-bad-test.js, but uses feed=longpoll, instead of feed=continuous
- Make sure changes-resource-leak-bad-test.js is not running
- IMPORTANT: fully stop both nodes. After *both* nodes have been stopped, start them back up. This is needed to “reset” the DB back to where we want it.
- $ node ./changes-resource-leak-good-test.js # This is will run indefinitely
- Even after many minutes you’ll see that both CouchDB nodes consume only a few percent of CPU each! Yippie!

## Notes:
- While changes-resource-leak-bad-test.js is running you can use `netstat` on both the server running changes-resource-leak-bad-test.js and on the CouchDB nodes to see that connections are being cleaned up and not persisting. Therefore, we can assume that changes-resource-leak-bad-test.js is not leaving open connections and that the leak is most likely in CouchDB itself.
- I throughly tested my load balancer to make sure that the load balancer was not causing the problem. I even used different load balancers.

**Comment by wohali** [comment]
@redgeoff I had a thought - can you try the fix in #1117 ? Just patch your local couchdb mochiweb with the patch from upstream and let me know if this works for you.

**Comment by redgeoff** [comment]
Sorry, I'm not sure I have the bandwidth to test this right now as it requires setting up a new test environment and it takes a bit of work. If I find some time I'll let you know.

**Comment by janl** [comment]
I finally had some time to do a reproducing setup. If anyone here wants to play with it, please contact me privately. I’m happy to hand out SSH access.

Big thanks to @redgeoff for proving such a comprehensive test case.

My results are preliminary, so nothing concluded yet.

Baseline info:
- I tested 2.1.1 and master as of today with Erlang R20.
- I test on DigitalOcean Droplets. Smallest one (1 “virtual“ CPU, 1.8GHz, 1GB RAM, SSD) and the second smallest “dedicated CPU” option (2 dedicated hyper threads, TBD GHz, 4GB RAM, SSD).
- I’m running haproxy on a. separate droplet.
- I’m running the node scripts on yet another droplet.

Observations:
- The differences I see are less in CPU time (things never go to 100%) but on load (fist digit).
  - On the smaller VM, I have one VM with occasional +1/3 load values.
  - This seems to correlate with a lager number of sockets on the system on `5984` in `TIME_WAIT` state.
- On the bigger machines, the difference is more like 1/4 or 1/5..
- On either nodes, I couldn’t really detect a sticky workhorse node, but the load differential swapping between the two CouchDB nodes. I didn’t detect a pattern, but (see graphs below), you see a slight drop in CPU % about once an hour.

Graphs from a ~3 hour run on the higher perf machines, this is CPU %:

Node 1:
![screen shot 2018-07-08 at 23 17 06](https://user-images.githubusercontent.com/11321/42424056-49337f00-8305-11e8-8471-bd1976b448cd.png)

Node 2:
![screen shot 2018-07-08 at 23 16 44](https://user-images.githubusercontent.com/11321/42424057-51dfa8fe-8305-11e8-9c31-58dc702391b1.png)

Things I wanna try next:
- run the long poll option to get a feel for what numbers to expect
- set `option http-close` in the haproxy config
- disable http keepalive on the CouchDB side (should be about the same effect as the previous item).

Gut feel: there is some variance in when TCP sockets get out of TIME_WAIT which accounts for the extra load and properly closing those HTTP requests either in the client or in the load balancer will help, or a reduction of the TIME_WAIT timeout, but not sure how well this is tunable across different systems.

This could be sockets being left in TIME_WAIT because of continuous changes requests not terminating with a. TCP FIN and/or unread bytes on the socket, and system-induced, or test-case induced variance in when those sockets are cleaned up, leading to larger load on the systems.

Might be worth re-reading https://vincent.bernat.im/en/blog/2014-tcp-time-wait-state-linux to refresh the memories.

So far for now. I’ll update this as I have more info.

**Comment by wohali** [comment]
@janl yes, reference #1417 for the scenario you described, which isn't scheduler-specific (just scheduler-exacerbated).

**Comment by janl** [comment]
One thing I left out in my last post:
- CouchDB _stats and _system do not show anything untoward, or askew with regards to the more busy node.

**Comment by janl** [comment]
Tried running the `good-test.js` on the smaller VMs and seeing no difference in CPU usage to `bad-test.js`. No difference in sockets in `TIME_WAIT` either. I’m inclined to say that this setup can not reproduce this.

**Comment by redgeoff** [comment]
Interesting. I was able to reproduce it multiple times and with complete success. I believe I created this issue in reference to CouchDB 2.1.0. Perhaps this issue was fixed in CouchDB 2.1.1?

The only other thing I can think of is that I used AWS for all my testing and perhaps the difference in network topology is somehow masking/creating the issue.

**Comment by wohali** [comment]
@redgeoff Were you perhaps using the `httpsd` daemon to connect to CouchDB over port 6894 instead of 5984? Erlang has some known...issues...with its TLS support.

If not, my guess is [this commit](https://github.com/apache/couchdb/pull/1173) fixed it, i.e. 302bd8b. @janl are you able to test on the commit prior, i.e. 4976f49 ?

**Comment by redgeoff** [comment]
I don’t believe I had TLS enabled for these tests.

**Comment by janl** [comment]
I can try 2.1.0.

@wohali this is all after 2.1.1, so not sure how it’s related.

**Comment by wohali** [comment]
@redgeoff We believe this is fixed in 2.2.0, are you able to confirm? Would like to close this ticket out.

**Comment by redgeoff** [comment]
Sorry, but I'm not going to have time to test this properly on AWS, so I'll just close this out. Thanks for working on it!
